MSQ generates tombstones honoring granularity specified in a `REPLACE` query.

abhishekrb19 opened the pull request:
Unlike native batch ingestion, MSQ is strict about segment allocation --  it will throw an `InsertCannotAllocateSegment` [fault](https://druid.apache.org/docs/latest/multi-stage-query/reference#error-codes) if the requested segment granularity  doesn't align with the allocated segment granularity. 

This patch modifies the segment generation logic in MSQ `REPLACE` to:
- Honor the segment granularity specified in the query for all finite-interval tombstones upto 5000 segments (`Limits.MAX_PARTITION_BUCKETS`) - e.g., `[2000, 2010)`. Once the limit exceeds, a `TooManyBucketsFault` fault is returned similar to data segments.
- For infinite-interval tombstones, retain the current behavior of generating one tombstone each for positive and negative eternity - e.g., `[-Inf, 2000)`, `[2010, Inf)`.

This behavior is consistent with how MSQ generally works re honoring segment granularity in the query. It's also similar to how native batch ingestion generates tombstones.


<!-- Thanks for trying to help us make Apache Druid be the best it can be! Please fill out as much of the following information as is possible (where relevant, and remove it when irrelevant) to help make the intention and scope of this PR clear in order to ease review. -->

<!-- Please read the doc for contribution (https://github.com/apache/druid/blob/master/CONTRIBUTING.md) before making this PR. Also, once you open a PR, please _avoid using force pushes and rebasing_ since these make it difficult for reviewers to see what you've changed in response to their reviews. See [the 'If your pull request shows conflicts with master' section](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master) for more details. -->


<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description
[Previously](https://github.com/apache/druid/pull/13893/) MSQ replace would generate "irregular" tombstones as an optimization so the controller doesn't run OOM. The new code only treats the infinite interval ones special, which will be cleaned up by the coordinator in a follow up change.

This change semi-reverts https://github.com/apache/druid/pull/13893/, uses the tests added in the PR and adds a new condition to account for the infinite-interval tombstones.

<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->


<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

#### Release note

MSQ `REPLACE` queries will generate tombstone segments honoring the segment granularity specified in the query. If a query generates more than 5000 tombstones, it'll return an MSQ `TooManyBucketsFault` to the user, similar to the behavior with data segments.
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->


<hr>

##### Key changed/added classes in this PR
 * `TombstoneHelper.java`
 * `TombstoneHelperTest.java`


<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

gianm reviewed (commented):
In general I like the idea of having the tombstone behavior be better aligned between native and MSQ. I'm not sure if there's a reason that it needs to be different.

Couple comments:

1. Looks like `TombstoneHelper` has `computeTombstones` used by native batch in replace mode, and `computeTombstoneSegmentsForReplace` used by MSQ in replace mode. Is there a practical difference in the logic? If so— is it a difference we need to preserve, or could both systems use the same method? It looks like the method used by MSQ was added in #13706. Perhaps the PR has some discussion that would shed light on these questions.
2. Ideally MSQ would return a `TooManyBuckets` fault if there are more than 5000 tombstones (`StageDefinition.PARTITION_STATS_MAX_BUCKETS`). To support this I suggest adding a limit parameter to `TombstoneHelper`, so there's some way for a caller to tell if the limit is exceeded. This prevents OOM in situations where there are too many time chunks. It also aligns with the behavior of an ingest that generates actual data as opposed to tombstones.

abhishekrb19 commented:
Thank you for the review, @gianm!

1. Yep, I agree. It looks like @LakshSingla did some analysis re reusing `computeTombstones`- https://github.com/apache/druid/pull/13706#discussion_r1093878894. I also looked into it, and I think with some refactoring like moving [IntervalUtils](https://github.com/apache/druid/blob/master/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/util/IntervalUtils.java) into core druid and passing in specific arguments, this should be possible. I'd prefer to revisit the refactor as a separate PR from the functional change in this patch. For now, I've added javadocs to the two methods the native batch and MSQ engines use. 
2. Good idea! I updated the code per your suggestion and added unit tests in `MSQReplaceTest.java` and `TombstoneHelperTest.java`.

LakshSingla commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java:
```diff
@@ -1423,14 +1424,18 @@ private void publishAllSegments(final Set<DataSegment> segments) throws IOExcept
         catch (IllegalStateException e) {
           throw new MSQException(e, InsertLockPreemptedFault.instance());
         }
+        catch (IllegalArgumentException e) {
+          throw new MSQException(e, new TooManyBucketsFault(Limits.MAX_PARTITION_BUCKETS));
+        }
```
I think this will also change, IAE doesn't seem valid here.

abhishekrb19 replied:
Please see my related comment related to throwing MSQ faults from the indexing package. Also, this follow the same pattern as `InsertLockPreemptedFault` above

LakshSingla replied:
We can pull out the exception above. 
`IAE` to me is more like entering a negative integer instead of a negative one, or a different type. The logic inside the helper operates before comparing it with the maxBuckets parameter. Or perhaps there are "XOR args" (enter exactly one) and the user passes both. 
As per InsertLockPreempted, it might be insufficiency at the time when it was written, but ISE seems more forgiving than IAE. Also, that logic in itself is also pretty fragile, if the logic switches to `DruidException`, then it would break.
At worst, IAE can have some special text inside, and we do a string compare before throwing this exception, however since we already have a special exception created, that seems like a valid path to follow

abhishekrb19 replied:
My bad, you were referring to `TooManyBucketsException` and not the MSQ fault `TooManyBucketsFault`. Moved the exception to the `indexing-service` package for better exception handling semantics.

LakshSingla reviewed (commented):
Since TooManyBucketsFault now also includes tombstones, does the user-facing message for the error still hold, or do we want it to be more explanatory?
Expressing "tombstones" in the message doesn't seem neat, so I think we can mention the number of buckets generated & the number of empty buckets in the message to help the user & admin debug faster, without saying what tombstones are.

abhishekrb19 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/TombstoneHelper.java:
```diff
@@ -173,66 +190,69 @@ public Set<DataSegment> computeTombstoneSegmentsForReplace(
           continue;
         }
+        if (buckets > maxBuckets) {
+          throw new IAE("Cannot add more tombstone buckets than [%d].", maxBuckets);
```
I considered just returning the `TooManyBucketsException` directly here. But since this utility resides is in the common indexing package, we cannot use an MSQ fault `TooManyBucketsException`  here. So it has to be a more generic exception or introduce other semantics.

abhishekrb19 commented on extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQReplaceTest.java:
```diff
@@ -985,13 +1041,61 @@ public void testReplaceTombstonesOverPartiallyOverlappingSegments()
   }
 
+  @Test
+  public void testReplaceTombstonesWithTooManyBucketsThrowsException()
```
yeah, good call! Will move the test

abhishekrb19 commented:
@LakshSingla re:
> Since TooManyBucketsFault now also includes tombstones, does the user-facing message for the error still hold, or do we want it to be more explanatory?


The user-facing error message would still be:
```
TooManyBuckets: Too many partition buckets (max = 5,000); try breaking your query up into smaller queries or using a wider PARTITIONED BY
```

Did you mean the cause in the stacktrace? If so, I think it's better to be direct and say "tombstone" in the cause since it's an existing concept in the Druid docs.

LakshSingla commented:
> Did you mean the cause in the stacktrace

I meant the user facing message. We can enhance the message to include more information about what the user is doing, and how many buckets they are generating.

> I think it's better to be direct and say "tombstone" in the cause since it's an existing concept explained in the Druid docs.

I guess so, though tombstones isn't something that we expect a user to be familiar with.

abhishekrb19 commented:
@LakshSingla, thank you for the review! Regarding mentioning the total # of buckets a query would generate without the max buckets limit, the computation could lead to an OOM.  The `max = 5,000` in the exception message provides a safe permissible upper bound limit to a user, which I think suffices. 

I believe I've addressed your comments. Can you please take another look?

LakshSingla commented on extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQFaultsTest.java:
```diff
@@ -403,6 +413,52 @@ public void testReplaceWithAppendAndSharedLocks()
+
+    // Create a datasegment which lies partially outside the generated segment
+    DataSegment existingDataSegment = DataSegment.builder()
+                                                 .interval(Intervals.of("2001-01-01T/2003-01-04T"))
```
Can you also add a test case where this is eternity, the overwrite clause is eternity and there is no WHERE clause.

LakshSingla commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/TombstoneHelper.java:
```diff
@@ -173,66 +188,66 @@ public Set<DataSegment> computeTombstoneSegmentsForReplace(
-        retVal.add(alignedTombstoneInterval);
+          // Helps in deduplication if required. Since all the intervals are uniformly granular, there should be no
+          // no overlap post deduplication
+          HashSet<Interval> intervals = Sets.newHashSet(intervalsToDropByGranularity.granularityIntervalsIterator());
```
This is problematic I assume. If the overlap is `ETERNITY`,  `Sets.newHashSet(intervalsToDropByGranularity.granularityIntervalsIterator)` would materialize all the buckets, which would OOM, unless there's a limit on the iterator, which I missed.
The iterator should be limited, or we should walk through the iterator and then pick up the items as we traverse.

I have added a comment about another test case, which I feel would OOM due to this line

abhishekrb19 replied:
yes, good point. Updated with tests in https://github.com/apache/druid/pull/15243/commits/5d65ce0ad3391faf3f870c9b88819e923800a715
